tcti-device: Copy response header to output buffer in receive function.

flihp opened the pull request:
This commit resolves a bug in the device TCTI where the receive function
read the response header from the TPM but neglects to copy this to the
buffer provided by the caller. The unit tests for this module were
unfortunately not catching this misbehavior.

For this fix we add a function to the tcti module to assist in
marshalling the response header back out to the caller provided buffer.
The unit test for the device TCTI has also been updated to compare the
response buffer received by the client to the pre-canned response buffer
that the mock framework uses when mocking the 'read' function.

Signed-off-by: Philip Tricca <philip.b.tricca@intel.com>

flihp commented:
this resolves the issue discussed in #879

tstruk commented on src/tss2-tcti/tcti-device.c:
```diff
@@ -131,9 +131,13 @@ tcti_device_receive (
                      *response_size, tcti_intel->header.size);
         return TSS2_TCTI_RC_INSUFFICIENT_BUFFER;
     }
+    rc = header_marshal (&tcti_intel->header, response_buffer);
```
Do we need the header_buf at all? We could just read the header into to the response_buffer and 
get the header from it: parse_header (response_buffer, &tcti_intel->header);
This way we would not need the header_marshal at all.

AndreasFuchsSIT replied:
Yes, because an application may call tcti_receive() with buffer==NULL.

flihp replied:
When this code was being refactored I tried both of these approaches. Unmarshalling and storing the header this way just looked better than unmarshalling the size field from a buffer when we needed it. If we're looking to be more efficient we could get away with using 6 instead of 10 bytes since we don't do anything with the RC but I didn't see any harm in storing the whole header.

martinezjavier commented:
@flihp while reading the code to review this, I noticed that another flaw in the[ tcti_device_receive()](https://github.com/tpm2-software/tpm2-tss/blob/master/src/tss2-tcti/tcti-device.c#L82) logic:

```
    *response_size = tcti_intel->header.size;
    if (response_buffer == NULL) {
        LOG_DEBUG ("response_buffer is null, returning size: %zd", *response_size);
        return TSS2_RC_SUCCESS;
    }
    if (*response_size < tcti_intel->header.size) {
        LOG_WARNING ("Size of user supplied response buffer %zd is less than "
                     "the size of the response buffer: %" PRIu32,
                     *response_size, tcti_intel->header.size);
        return TSS2_TCTI_RC_INSUFFICIENT_BUFFER;
    }
```

I believe you want to set ` *response_size = tcti_intel->header.size` after you have checked for `*response_size < tcti_intel->header.size`.

I also think that found an issue in [read_all()](https://github.com/tpm2-software/tpm2-tss/blob/master/src/util/io.c#L40):

```
    do {
        recvd = TEMP_RETRY (read (fd, &data [recvd_total], size));
        if (recvd < 0) {
            LOG_WARNING ("read on fd %d failed with errno %d: %s",
                         fd, errno, strerror (errno));
            return recvd_total;
        }
        LOGBLOB_DEBUG (&data [recvd_total], recvd, "read %zd bytes from fd %d:", recvd, fd);
        recvd_total += recvd;
        size -= recvd;
    } while (size > 0);
```

If read(2) returns 0, then this will be an infinite loop. Probably that will be a bug, but still is a case that should be handled.

Maybe you can fix those as part of this pull request? Or if you want I can propose fixes as a follow-up.

flihp commented:
thanks @martinezjavier. The first issue you point out is one that I introduced in this patch. Thought I was being clever but I missed the issue you noted. Looks like another unit test waiting to be written. The second issue is on the read path and wasn't introduced here (I hope) so that should probably get it's own issue and PR.

martinezjavier commented:
@flihp both issues are not related to your patch in fact, and present in current master. I noticed when reading the current code to understand what your change was about. So mentioned you wanted to do some preparatory patches since you were already touching that part of the code.

But as mentioned those can be done as a follow-up, I can write patches for these after your changes are merged.

flihp commented:
You're right. I remember moving around the code that deals with the response size but I must have moved it back. I'll sort that one out here since we've already got the context for it. Will be in a second commit though.

flihp commented:
Based on review comments I've added a new commit that fixes another bug identified by @martinezjavier above in the setting of the `response_size` parameter in the `receive` function. This commit also adds a test case to ensure the proper behavior is preserved. Additionally I've renamed the `parse_header` function so that it's consistent with the `header_marshal` function added in the first commit.
